File: 0740-Delete-And-Earn/0740.java
class Solution {
  public int deleteAndEarn(int[] nums) {
    int maxVal = 0;
    for (int val : nums) {
      maxVal = Math.max(maxVal, val);
    }
    int[] sum = new int[maxVal + 1];
    for (int val : nums) {
      sum[val] += val;
    }
    return rob(sum);
  }

  public int rob(int[] nums) {
    int size = nums.length;
    int first = nums[0], second = Math.max(nums[0], nums[1]);
    for (int i = 2; i < size; i++) {
      int temp = second;
      second = Math.max(first + nums[i], second);
      first = temp;
    }
    return second;
  }
}

File: 0740-Delete-And-Earn/0740.py
class Solution:
  def deleteAndEarn(self, nums: List[int]) -> int:
    maxVal = max(nums)
    total = [0] * (maxVal + 1)
    for val in nums:
      total[val] += val

    def rob(nums: List[int]) -> int:
      size = len(nums)
      first, second = nums[0], max(nums[0], nums[1])
      for i in range(2, size):
        first, second = second, max(first + nums[i], second)
      return second

    return rob(total)